File: src/tailscale.test.ts
import { describe, expect, test } from 'bun:test';
import { decideServeAction, ensureTailscaleServe, parseServeProxyPort } from './tailscale';

const statusJson = (proxy: string | null) =>
  JSON.stringify({
    TCP: { '443': { HTTPS: true } },
    Web: {
      'host.tail0.ts.net:443': { Handlers: proxy === null ? {} : { '/': { Proxy: proxy } } },
    },
  });

describe('parseServeProxyPort', () => {
  test('루트 핸들러의 루프백 프록시 포트를 뽑는다', () => {
    expect(parseServeProxyPort(statusJson('http://127.0.0.1:8636'))).toBe(8636);
    expect(parseServeProxyPort(statusJson('http://localhost:8995'))).toBe(8995);
  });

  test('serve 미설정 / 파싱 불가 / 루트 아님은 null', () => {
    expect(parseServeProxyPort('{}')).toBeNull();
    expect(parseServeProxyPort('')).toBeNull();
    expect(parseServeProxyPort('No serve config')).toBeNull();
    expect(parseServeProxyPort(statusJson(null))).toBeNull();
  });

  test('루프백이 아닌 프록시 대상은 우리 관심사가 아니다', () => {
    expect(parseServeProxyPort(statusJson('http://192.168.0.5:8636'))).toBeNull();
  });

  test('루트가 아닌 경로만 있으면 null', () => {
    const json = JSON.stringify({
      Web: { 'host:443': { Handlers: { '/other': { Proxy: 'http://127.0.0.1:8636' } } } },
    });
    expect(parseServeProxyPort(json)).toBeNull();
  });
});

describe('decideServeAction', () => {
  test('빈 자리면 잡는다', () => {
    expect(
      decideServeAction({ myPort: 8636, occupiedPort: null, occupantIsLiveDaemon: false }).action,
    ).toBe('claim');
  });

  test('이미 내 포트면 no-op', () => {
    expect(
      decideServeAction({ myPort: 8636, occupiedPort: 8636, occupantIsLiveDaemon: false }).action,
    ).toBe('keep');
  });

  test('살아있는 다른 데몬이 쓰는 중이면 양보한다', () => {
    const decision = decideServeAction({
      myPort: 8995,
      occupiedPort: 8636,
      occupantIsLiveDaemon: true,
    });
    expect(decision.action).toBe('yield');
    expect(decision.message).toContain('8636');
  });

  test('죽은 포트를 가리키면 되찾는다', () => {
    const decision = decideServeAction({
      myPort: 8636,
      occupiedPort: 8995,
      occupantIsLiveDaemon: false,
    });
    expect(decision.action).toBe('reclaim');
  });
});

/** run 호출을 기록하는 스텁 — serve 하위 명령이 실제로 실행됐는지 본다. */
function stubRun(statusOut: string, statusOk = true) {
  const calls: string[][] = [];
  const run = (args: string[]) => {
    calls.push(args);
    return args[1] === 'status' ? { ok: statusOk, out: statusOut } : { ok: true, out: 'started' };
  };
  const served = () => calls.some((args) => args.includes('--bg'));
  return { run, calls, served };
}

describe('ensureTailscaleServe', () => {
  test('살아있는 다른 데몬의 노출은 빼앗지 않는다', async () => {
    const stub = stubRun(statusJson('http://127.0.0.1:8636'));
    const log = await ensureTailscaleServe(8995, {
      run: stub.run,
      probeDaemon: async () => true,
    });
    expect(stub.served()).toBe(false);
    expect(log).toContain('건드리지 않는다');
  });

  test('stale 설정(죽은 포트)은 되찾는다', async () => {
    const stub = stubRun(statusJson('http://127.0.0.1:8995'));
    await ensureTailscaleServe(8636, { run: stub.run, probeDaemon: async () => false });
    expect(stub.calls).toContainEqual(['serve', '--bg', '8636']);
  });

  test('이미 내 포트면 probe 없이 no-op', async () => {
    const stub = stubRun(statusJson('http://127.0.0.1:8636'));
    let probed = false;
    await ensureTailscaleServe(8636, {
      run: stub.run,
      probeDaemon: async () => {
        probed = true;
        return true;
      },
    });
    expect(stub.served()).toBe(false);
    expect(probed).toBe(false);
  });

  test('serve 미설정이면 잡는다', async () => {
    const stub = stubRun('{}');
    await ensureTailscaleServe(8636, { run: stub.run, probeDaemon: async () => false });
    expect(stub.calls).toContainEqual(['serve', '--bg', '8636']);
  });

  test('status 조회 실패는 fail-open — claim 을 시도하고 실패해도 던지지 않는다', async () => {
    const calls: string[][] = [];
    const run = (args: string[]) => {
      calls.push(args);
      return { ok: false, out: 'tailscale CLI 를 찾을 수 없다' };
    };
    const log = await ensureTailscaleServe(8636, { run, probeDaemon: async () => false });
    expect(calls).toContainEqual(['serve', '--bg', '8636']);
    expect(log).toContain('무시하고 계속');
  });
});

File: crates/rocky-todod/tests/tailscale_test.rs
//! TS `src/tailscale.test.ts` 포팅.

use std::sync::{Arc, Mutex};

use rocky_todod::runner::{CmdOutput, Runner};
use rocky_todod::tailscale::*;

fn json_for(port: u16) -> String {
    format!(
        r#"{{"Web":{{"host.ts.net:443":{{"Handlers":{{"/":{{"Proxy":"http://127.0.0.1:{port}"}}}}}}}}}}"#
    )
}

#[test]
fn parses_loopback_proxy_port_at_root() {
    assert_eq!(parse_serve_proxy_port(&json_for(8636)), Some(8636));
}

#[test]
fn ipv6_loopback_counts() {
    let json = r#"{"Web":{"h:443":{"Handlers":{"/":{"Proxy":"http://[::1]:8993"}}}}}"#;
    assert_eq!(parse_serve_proxy_port(json), Some(8993));
    let localhost = r#"{"Web":{"h:443":{"Handlers":{"/":{"Proxy":"http://localhost:8993"}}}}}"#;
    assert_eq!(parse_serve_proxy_port(localhost), Some(8993));
}

#[test]
fn unset_or_unparsable_is_none() {
    assert_eq!(parse_serve_proxy_port(""), None);
    assert_eq!(parse_serve_proxy_port("not json"), None);
    assert_eq!(parse_serve_proxy_port("{}"), None);
    assert_eq!(parse_serve_proxy_port(r#"{"Web":{}}"#), None);
}

#[test]
fn non_loopback_target_is_ignored() {
    let json = r#"{"Web":{"h:443":{"Handlers":{"/":{"Proxy":"http://192.168.1.5:8636"}}}}}"#;
    assert_eq!(parse_serve_proxy_port(json), None);
}

#[test]
fn non_root_paths_only_is_none() {
    let json = r#"{"Web":{"h:443":{"Handlers":{"/api":{"Proxy":"http://127.0.0.1:8636"}}}}}"#;
    assert_eq!(parse_serve_proxy_port(json), None);
}

#[test]
fn empty_slot_claims() {
    let decision = decide_serve_action(8636, None, false);
    assert_eq!(decision.action, ServeAction::Claim);
}

#[test]
fn own_port_keeps() {
    let decision = decide_serve_action(8636, Some(8636), false);
    assert_eq!(decision.action, ServeAction::Keep);
}

#[test]
fn live_other_daemon_yields() {
    let decision = decide_serve_action(8993, Some(8636), true);
    assert_eq!(decision.action, ServeAction::Yield);
    assert!(decision.message.contains(":8636"));
}

#[test]
fn dead_port_reclaims() {
    let decision = decide_serve_action(8636, Some(8993), false);
    assert_eq!(decision.action, ServeAction::Reclaim);
}

// ── ensure_tailscale_serve ──

fn recording_runner(status_json: String) -> (Runner, Arc<Mutex<Vec<Vec<String>>>>) {
    let calls: Arc<Mutex<Vec<Vec<String>>>> = Arc::new(Mutex::new(Vec::new()));
    let sink = calls.clone();
    let runner: Runner = Arc::new(move |cmd, _stdin, _timeout| {
        sink.lock().unwrap().push(cmd.clone());
        let out = if cmd.get(2).map(String::as_str) == Some("status") {
            CmdOutput {
                code: 0,
                stdout: status_json.clone(),
                stderr: String::new(),
            }
        } else {
            CmdOutput {
                code: 0,
                stdout: "serve set".into(),
                stderr: String::new(),
            }
        };
        Box::pin(async move { out })
    });
    (runner, calls)
}

fn probe(alive: bool) -> impl Fn(u16) -> rocky_todod::runner::BoxFut<bool> {
    move |_port| Box::pin(async move { alive })
}

#[tokio::test]
async fn does_not_steal_live_daemons_serve() {
    let (runner, calls) = recording_runner(json_for(8636));
    let message = ensure_tailscale_serve(&runner, 8993, probe(true)).await;
    assert!(message.contains("건드리지 않는다"), "{message}");
    // serve --bg 를 실행하지 않았다
    let calls = calls.lock().unwrap();
    assert!(calls
        .iter()
        .all(|c| c.get(2).map(String::as_str) != Some("--bg")));
    assert!(!calls.iter().any(|c| c.contains(&"--bg".to_string())));
}

#[tokio::test]
async fn stale_dead_port_is_reclaimed() {
    let (runner, calls) = recording_runner(json_for(8636));
    let message = ensure_tailscale_serve(&runner, 8993, probe(false)).await;
    assert!(message.contains("되돌린다"), "{message}");
    let calls = calls.lock().unwrap();
    assert!(calls.iter().any(|c| c.contains(&"--bg".to_string())));
}

#[tokio::test]
async fn own_port_is_noop_without_probe() {
    let (runner, calls) = recording_runner(json_for(8636));
    let probed = Arc::new(Mutex::new(false));
    let flag = probed.clone();
    let message = ensure_tailscale_serve(&runner, 8636, move |_| {
        *flag.lock().unwrap() = true;
        Box::pin(async { true })
    })
    .await;
    assert!(message.contains("이미"), "{message}");
    assert!(!*probed.lock().unwrap());
    let calls = calls.lock().unwrap();
    assert!(!calls.iter().any(|c| c.contains(&"--bg".to_string())));
}

#[tokio::test]
async fn unset_serve_claims() {
    let (runner, calls) = recording_runner("{}".into());
    let message = ensure_tailscale_serve(&runner, 8636, probe(false)).await;
    assert!(message.contains("활성화"), "{message}");
    let calls = calls.lock().unwrap();
    assert!(calls.iter().any(|c| c.contains(&"--bg".to_string())));
}

#[tokio::test]
async fn status_failure_is_fail_open() {
    let runner: Runner = Arc::new(|cmd, _stdin, _timeout| {
        Box::pin(async move {
            if cmd.get(2).map(String::as_str) == Some("status") {
                CmdOutput::failure("boom")
            } else {
                CmdOutput {
                    code: 1,
                    stdout: String::new(),
                    stderr: "serve failed".into(),
                }
            }
        })
    });
    let message = ensure_tailscale_serve(&runner, 8636, probe(false)).await;
    // claim 을 시도하고 실패해도 던지지 않는다
    assert!(message.contains("실패"), "{message}");
}
